### PR TITLE
Support patches that create or delete datasets

### PR DESCRIPTION
**Pull request by olsen232**
## Related links:

Fixes https://github.com/koordinates/sno/issues/239

## Checklist:

- [x] Have you reviewed your own change?
- [x] Have you included test(s)?
- [x] Have you updated the [changelog](https://github.com/koordinates/sno/blob/master/CHANGELOG.md)?

### CODE REVIEW

**Review by craigds — APPROVED** [review]
LGTM

File: sno/dataset2.py [hunk 1]
```diff
@@ -401,9 +395,11 @@ def apply_meta_diff(self, meta_diff, tree_builder):
         if not meta_diff:
             return
 
-        conflicts = False
+        # Applying diffs works even if there is no tree yet created for the dataset,
+        # as is the case when the dataset is first being created right now.
+        meta_tree = self.meta_tree if self.tree is not None else ()
 
-        meta_tree = self.meta_tree
+        conflicts = False
```
Comment (rcoup) [inline sno/dataset2.py:402]: would prefer to use `has_`/`is_`/`can_`/etc as a prefix for boolean variable names

File: sno/git_util.py [hunk 2]
```diff
@@ -6,6 +6,38 @@
 from .timestamps import tz_offset_to_minutes
 
 
+def get_head_tree(repo):
+    """Returns the tree at the current repo HEAD."""
```
Comment (rcoup) [inline sno/git_util.py:10]: Can the docstring be expanded with the comments about when/why `None` is returned?

File: sno/git_util.py [hunk 3]
```diff
@@ -6,6 +6,38 @@
...
+        return None
+    try:
+        return repo.head.peel(pygit2.Tree)
+    except pygit2.GitError:
+        # This happens when the repo is not empty, but the current HEAD has no commits.
+        return None
+
+
+def get_head_commit(repo):
+    """Returns the commit at the current repo HEAD."""
```
Comment (rcoup) [inline sno/git_util.py:21]: Can the docstring be expanded with the comments about when/why `None` is returned?

File: sno/rich_tree_builder.py [hunk 4]
```diff
@@ -13,18 +16,27 @@ class RichTreeBuilder:
     Conflicts are not detected.
     """
 
-    def __init__(self, repo, initial_root_tree):
+    def __init__(self, repo, initial_root_tree, auto_include_version_blobs=True):
         """
         The repo and an initial root tree which will be updated.
         All paths are specified relative to this tree - the root tree at a particular commit is a good choice.
```
Comment (rcoup) [inline sno/rich_tree_builder.py:22]: document what `auto_include_version_blobs` does, since it's not obvious?

**Review by rcoup — COMMENTED** [review]
Nice, some minor doc/style comments

File: sno/dataset2.py (hunk 1 above)
Comment (olsen232) [reply sno/dataset2.py:402]: Done

File: sno/git_util.py (hunk 2 above)
Comment (olsen232) [reply sno/git_util.py:10]: Done. Also fixed the try/except pattern - we should have just been checking if head_is_unborn

File: sno/git_util.py (hunk 3 above)
Comment (olsen232) [reply sno/git_util.py:21]: Done

File: sno/rich_tree_builder.py (hunk 4 above)
Comment (olsen232) [reply sno/rich_tree_builder.py:22]: Actually, this should be at a different level - the tree builder should just build the tree its told to build and not be clever about it.
I've moved it to the caller - RepositoryStructure.create_tree_from_diff - and added a comment

**Pull request merged**